src/transfers/monitor.hpp

```diff
@@ -0,0 +1,259 @@
+        /// bothered to call the done method there.
+        Dropped,
+    };
+
+    /// A Status of a running transfer.
+    class Status {
+    private:
+        friend class Monitor;
+        Lock lock;
+        Status(Lock &&lock)
+            : lock(std::move(lock)) {}
+
+    public:
+        Status(Status &&other) = default;
+        Status &operator=(Status &&other) = default;
+        Status(const Status &other) = delete;
+        Status &operator=(const Status &other) = delete;
+        Type type;
+        TransferId id;
+
+        /// When the transfer started.
+        ///
+        /// In our own internal timestamp ‒ not necessarily anchored to
+        /// anything and may wrap around!
+        ///
+        /// In milliseconds.
+        Timestamp start;
+
+        /// The expected size.
+        size_t expected;
+
+        /// How much got already transferred.
+        size_t transferred;
+
+        /// The path where the transfer will be stored (once it's done).
+        ///
+        /// Note that usually the file doesn't yet exist, the data is being
+        /// saved into some kind of temp file and then moved.
+        const char *destination;
+    };
+
+    /// An allocated transfer slot.
+    ///
+    /// Keep alive while the transfer is running.
+    class Slot {
+    private:
+        friend class Monitor;
+        Monitor &owner;
+        Slot(Monitor &owner);
+        // Set to false when moved from.
+        bool live;
+
+    public:
+        Slot(Slot &&other);
+        Slot &operator=(Slot &&other) = delete;
+        Slot(const Slot &other) = delete;
+        Slot &operator=(const Slot &other) = delete;
+        ~Slot();
+        /// Update the progress report.
+        ///
+        /// More bytes were transferred, account for them in the tracking.
+        ///
+        /// This is the increment, not the accumulated total.
+        void progress(size_t add_bytes);
+
+        /// The transfer is done.
+        ///
+        /// This reports the outcome of this transfer. The Slot shall not be
+        /// used more and shall be dropped consequently.
+        void done(Outcome reason);
+    };
+
+    friend class Slot;
+
+private:
+    // Mutex for the "active transfer" information.
+    mutable Mutex main_mutex;
+    // Mutex for the history of outcomes.
+    mutable Mutex history_mutex;
+    // Changes to these _are_ protected by a mutex, but it is possible to
+    // _read_ them without locking.
+    //
+    // A bit of care needs to be employed when dealing with them, to avoid
+    // races, etc.
+    std::atomic<bool> transfer_active = false;
+    std::atomic<TransferId> current_id;
+
+    // Transfer related
+    bool used = false;
+    Type type;
+    Timestamp start;
+    size_t expected;
+    size_t transferred;
+    char destination_path[FILE_PATH_BUFFER_LEN];
+
+    // History related
+    static constexpr size_t HISTORY_MAX_LEN = 2;
+    size_t history_len = 0;
+    TransferId history_end;
```

Comment:
This is probably a matter of personal taste, but `history_end` sounds like the oldest thing not the latest to me (also kind of like an end iterator for some array), IMHO something like `previous_id` or `latest_id` would be more clear, but I leave that up to you, if you want to change it or not, I'm fine with it if you think it's better this way.